[FLINK-17236][docs] Add Tutorials section overview (#11826)
apache/flink: docs/tutorials/index.md

writing scalable streaming ETL, analytics, and event-driven applications, while leaving out a lot of
(ultimately important) details. The focus is on providing straightforward introductions to Flink's
APIs for managing state and time, with the expectation that having mastered these fundamentals,
you'll be much better equipped to pick up the rest of what you need to know from the more detailed
reference documentation. The links at the end of each page will lead you to where you can learn
more.

Specifically, you will learn:

- how to implement streaming data processing pipelines
- how and why Flink manages state
- how to use event time to consistently compute accurate analytics
- how to build event-driven applications on continuous streams
- how Flink is able to provide fault-tolerant, stateful stream processing with exactly-once semantics

These tutorials focus on four critical concepts: continuous processing of streaming data, event
time, stateful stream processing, and state snapshots. This page introduces these concepts.

{% info Note %} Accompanying these tutorials are a set of hands-on exercises that will guide you

> through learning how to work with the concepts being presented.

Reviewer: add link to http://github.com/apache/flink-training ?

Reviewer: or mention that links to the according exercises will be available where needed?

Author: Done.

{% top %}

## Stream Processing

> Streams are data's natural habitat. Whether it's events from web servers, trades from a stock

Reviewer: nit: I usually try to spell out things like "it is" to simplify things for non-native speakers...
```suggestion
Streams are data's natural habitat. Whether it is events from web servers, trades from a stock
```

exchange, or sensor readings from a machine on a factory floor, data is created as part of a stream.
But when you analyze data, you can either organize your processing around _bounded_ or _unbounded_
streams, and which of these paradigms you choose has profound consequences.

<img src="{{ site.baseurl }}/fig/bounded-unbounded.png" alt="Bounded and unbounded streams" class="offset" width="90%" />

**Batch processing** is the paradigm at work when you process a bounded data stream. In this mode of
operation you can choose to ingest the entire dataset before producing any results, which means that

> it's possible, for example, to sort the data, compute global statistics, or produce a final report

Reviewer: ```suggestion
it is possible, for example, to sort the data, compute global statistics, or produce a final report
```

that summarizes all of the input.

**Stream processing**, on the other hand, involves unbounded data streams. Conceptually, at least,
the input may never end, and so you are forced to continuously process the data as it arrives. 

In Flink, applications are composed of **streaming dataflows** that may be transformed by
user-defined **operators**. These dataflows form directed graphs that start with one or more

> **sources**, and end in one or more **sinks**.

Reviewer: You could actually link to the glossary here, but I'm not sure whether it would confuse people if they actually followed the links and read the information there (may be too much detail there); however, it may be useful later.

Author: I'm trying to minimize links within each page, to encourage folks to continue reading until the end. And yes, I think going to the glossary this early would be a mistake.

> <img src="{{ site.baseurl }}/fig/program_dataflow.svg" alt="A DataStream program, and its dataflow." class="offset" width="80%" />

Reviewer: This code example should be updated:
* not using `keyBy("id")` based on a string - bad practice and may also be removed in the future
* not using the `BucketingSink` (this line also has a rendering error)

Author: I fixed the bucketing sink rendering error in https://github.com/apache/flink/pull/11828#pullrequestreview-396512480 -- this figure is already in the docs, I'm just reusing it. I do agree about keyBy("id") though.

Author: Once github.com/apache/flink/pull/11828 is merged then perhaps I can fix the keyBy issue in this figure.

Reviewer: since this is not a new figure, let's fix the use of the `BucketingSink` afterwards...

> Often there is a one-to-one correspondence between the transformations in the programs and the

Reviewer: ```suggestion
Often there is a one-to-one correspondence between the transformations in the program and the
```

operators in the dataflow. Sometimes, however, one transformation may consist of multiple operators.

An application may consume real-time data from streaming sources such as message queues or

> distributed logs, such as Apache Kafka or Kinesis. But flink can also consume bounded, historic data

Reviewer: don't repeat "such as" again? Maybe replace with this?
```suggestion
distributed logs, like Apache Kafka or Kinesis. But flink can also consume bounded, historic data
```

from a variety of data sources. Similarly, the streams of results being produced by a Flink
application can be sent to a wide variety of systems, and the state held within Flink can be

> accessed via a REST API.

Reviewer: Accessing state within Flink via a REST API? That's not really available, is it?

Author: Given the questionable nature of queryable state, I think I'll just drop this phrase.

<img src="{{ site.baseurl }}/fig/flink-application-sources-sinks.png" alt="Flink application with sources and sinks" class="offset" width="90%" />

### Parallel Dataflows

Programs in Flink are inherently parallel and distributed. During execution, a
*stream* has one or more **stream partitions**, and each *operator* has one or

> more **operator subtasks**. The operator subtasks are independent of one

Reviewer: the new term for "subtask" is "task", but, according to the [glossary](https://ci.apache.org/projects/flink/flink-docs-master/concepts/glossary.html#sub-task), a "Sub-Task" is the same but "emphasizes that there are multiple parallel Tasks for the same Operator or Operator Chain". I think, we should use "Task" here (not sure about capitalization).

Author: But here we are, in fact, emphasizing the parallel nature of the many sub-tasks for each operator. If we don't use the term here, then where would we use it?

Reviewer: Yes, you are right - should it be consistently "sub-task" as in the glossary? (I would also be fine with "subtask" here since it improves the flow)

another, and execute in different threads and possibly on different machines or
containers.

The number of operator subtasks is the **parallelism** of that particular

> operator. The parallelism of a stream is always that of its producing operator.

Reviewer: Do you actually need the concept of the "parallelism of a stream"? Because since there could be n->m streams, I find it difficult to just use "n" here and also I rarely use parallelism for the stream itself...

Author: Agreed. That sentence (about the parallelism of a stream) is just confusing. (Didn't write it, just moved it here from a page in the concepts section.) Dropping it.

Different operators of the same program may have different levels of
parallelism.

> <img src="{{ site.baseurl }}/fig/parallel_dataflow.svg" alt="A parallel dataflow" class="offset" width="80%" />

Reviewer: FYI: task vs. subtask should also be changed in the image if changed in text (as proposed above)

recorded in the data stream, rather than using the clocks of the machines processing the data.

{% top %}

## Stateful Stream Processing

Flink's operations can be stateful. This means that how one event is handled can depend on the
accumulated effect of all the events that came before it. State may be used for something simple,
such as counting events per minute to display on a dashboard, or for something more complex, such as
computing features for a fraud detection model.

A Flink application is run in parallel on a distributed cluster. The various parallel instances of a
given operator will execute independently, in separate threads, and in general will be running on
different machines.

The set of parallel instances of a stateful operator is effectively a sharded key-value store. Each
parallel instance is responsible for handling events for a specific group of keys, and the state for
those keys is kept locally.

The diagram below shows a job running with a parallelism of two across the first three operators in
the job graph, terminating in a sink that has a parallelism of one. The third operator is stateful,

> and you can see that a fully connected network shuffle is occurring between the second and third

Reviewer: ?
```suggestion
and you can see that a fully-connected network shuffle is occurring between the second and third
```

Author: Done.

operators. This is being done to partition the stream by some key, so that all of the events that
need to be processed together, will be.

<img src="{{ site.baseurl }}/fig/parallel-job.png" alt="State is sharded" class="offset" width="65%" />

State is always accessed locally, which helps Flink applications achieve high throughput and

> low-latency. You can choose to keep state on the JVM heap, or if it is too large, in efficiently
> organized on-disk data structures. 

Reviewer: ```suggestion
low-latency. You can choose to keep state on the JVM heap, or if it is too large, in
efficiently-organized on-disk data structures. 
```

Author: No, I think in this case the hyphen goes too far.

Author: Double-checked with my wife, the editor. She says no hyphen in this case.

Reviewer: ok, thanks - I'm always confused with that in English...even though the "efficiently" relates to "organized" and not "data structures"?
Examples how I would derive that otherwise:
- [efficient] [on-disk] data structures
- efficiently-organized [on-disk] data structures

<img src="{{ site.baseurl }}/fig/local-state.png" alt="State is local" class="offset" width="90%" />

{% top %}

## Fault Tolerance via State Snapshots

Flink is able to provide fault-tolerant, exactly-once semantics through a combination of state
snapshots and stream replay. These snapshots capture the entire state of the distributed pipeline,
recording offsets into the input queues as well as the state throughout the job graph that has
resulted from having ingested the data up to that point. When a failure occurs, the sources are
rewound, the state is restored, and processing is resumed. As depicted above, these state snapshots

> are captured asynchronously, without impeding the ongoing processing.

Reviewer: "Understatement of the year" ;)
But as for the training and as an introduction, it is fine to keep away the details here.

Author: Lol.
